aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -51,10 +55,73 @@ void THTensor_(geometric)(THTensor *self, THGenerator *_generator, double p)
   TH_TENSOR_APPLY(real, self, *self_data = (real)THRandom_geometric(_generator, p););
 }
 
+#ifdef TH_BLAS_MKL
+#define BERNOULLI_OMP 10000
+void Bernoulli_generate(const int64_t n, const double p, real *r, THGenerator *_generator)
+{
+  int64_t seed = 17 + THRandom_random(_generator) % 4096;
+
+#ifdef TH_REAL_IS_INT
+  int *tmp = r;
+#else
+  int *tmp = (int*)malloc(n*sizeof(int));
+#endif
+
+#ifdef _OPENMP
+  size_t nthr = omp_get_num_threads();
+  int run_parallel = 0;
+  if ((omp_in_parallel() == 0) && (n >= BERNOULLI_OMP)) {
+    run_parallel = 1;
+  }
+  if (0 == run_parallel) nthr = 1;
+
+#pragma omp parallel num_threads(nthr) firstprivate(nthr)
```

Comment:
Why is this `firstprivate` needed?

Reply:
All threads have their own variables will be more efficient that those share the same variable.